lib/dns.js

```diff
@@ -289,31 +289,38 @@ function setServers(servers) {
   // servers cares won't have any servers available for resolution
   const orig = this._handle.getServers();
   const newSet = [];
-  const IPv6RE = /\[(.*)\]/;
   const addrSplitRE = /(^.+?)(?::(\d+))?$/;
 
   servers.forEach((serv) => {
+    // Easily recognized?
     var ipVersion = isIP(serv);
     if (ipVersion !== 0)
       return newSet.push([ipVersion, serv, IANA_DNS_PORT]);
 
-    const match = serv.match(IPv6RE);
-    // we have an IPv6 in brackets
-    if (match) {
-      ipVersion = isIP(match[1]);
+    // IPv6: /\[(.*)\]/ but O(n) complexity
+    const openBracketIx = serv.indexOf('[');
+    const nextCloseBracketIx = serv.indexOf(']', openBracketIx);
+    if (0 <= openBracketIx && openBracketIx < nextCloseBracketIx) {
+      const withinBrackets = serv.slice(openBracketIx + 1, nextCloseBracketIx);
+      ipVersion = isIP(withinBrackets);
```

Comment:
Can't you match using `/\[([^\]]+)\]/`?  Succinct and has linear complexity... although it does look as if the cat sat on your keyboard.

Reply:
Actually, `/\[[^\]]+\]/` is still quadratic under input like `'['.repeat(N) + 'a'`. Lacking a leading `^` anchor, the regex engine will try each of the `[`s as a possible start point and will take linear time to reject each one.

Something like `/\[([^\[\]]+)\]/` would be safe but:
1. The cat's revenge is complete, and
2. It couldn't handle nested braces like `[[]]` (not sure if that's relevant in the DNS spec).

See also: [vuln-regex-detector #49](https://github.com/davisjam/vuln-regex-detector/pull/49).

Reply:
Right, I mean the anchored version.  Nested square brackets shouldn't be an issue, an IPv6 address only has the two.